### PR TITLE
EOS-22704: Re-Analyze small objects Read throughput with MD cache patch

### PR DESCRIPTION
**Pull request by imvenkip**
Client(s3server) will store the pool version and layout type if the  MO_ENF_META flag is set while index creation. Motr will store the pool version and layout type if the M0_ENF_META flag is not set while index creation.

As per initial design, we have to add M0_ENF_META flag in to object index flags. Later it was changed to entity flags to support MD and DATA path. It was found while re-analyzing small object performance.

Modified M0_ENF_META flag validation to improve small object performance.

Signed-off-by: Venkateswarlu Payidimarry <venkateswarlu.payidimarry@seagate.com>

### CODE REVIEW

**Review by madhavemuri — APPROVED** [review]

**Review by yatin-mahajan — APPROVED** [review]
Looks good to me.

File: motr/idx.c [hunk 1]
```diff
@@ -227,7 +227,7 @@ static int idx_op_init(struct m0_idx *idx, int opcode,
 		oi->oi_dtx = NULL;
 
 	if (opcode == M0_EO_CREATE && entity->en_type == M0_ET_IDX &&
-	    oi->oi_flags & M0_ENF_META) {
+	    entity->en_flags & M0_ENF_META) {
```
Comment (yeshpal-jain-seagate) [inline motr/idx.c:230]: better to use M0_IN instead of check for more readability

Comment (imvenkip) [reply motr/idx.c:230]: But M0_IN() used for between two values of one variable. Here we are checking on value for opcode. Still i didn't get it. Can you explain more ?
Generally we are using M0_IN like
M0_IN(op->op_code, (M0_EO_CREATE, M0_EO_DELETE)

**Comment by andriytk** [comment]
Could you please elaborate - why your change improves performance and on how much (if there are any measurements)?

File: motr/idx.c (hunk 1 above)
Comment (yeshpal-jain-seagate) [reply motr/idx.c:230]: Right, Lets go with existing method, we do not have any Macro to check for single parameter.

**Comment by imvenkip** [comment]
> Could you please elaborate - why your change improves performance and on how much (if there are any measurements)?

Added M0_ENF_META flag to identify metadata. Client(s3server) will store the pool version and layout type if the MO_ENF_META flag is set while index creation. Motr will store the pool version and layout type if the
M0_ENF_META flag is not set while index creation. Please check reference EOS-23040. 

EOS-22704 is addressing from Motr and S3. I collected stats with 1726 main build and with motor changes.
Please check in JIRA link to find small obj read performance improvement stats.

https://jts.seagate.com/browse/EOS-22704?focusedCommentId=2176626&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-2176626

**Review by yeshpal-jain-seagate — APPROVED** [review]

**Comment by andriytk** [comment]
People from the Open Source community don't have access to our Jira tickets. cc @johnbent

**Comment by ivan-alekhin** [comment]
>> Could you please elaborate - why your change improves performance and on how much (if there are any measurements)?

> Added M0_ENF_META flag to identify metadata. Client(s3server) will store the pool version and layout type if the MO_ENF_META flag is set while index creation. Motr will store the pool version and layout type if the
M0_ENF_META flag is not set while index creation. Please check reference EOS-23040.

> EOS-22704 is addressing from Motr and S3. I collected stats with 1726 main build and with motor changes.
Please check in JIRA link to find small obj read performance improvement stats.

It is still not clear how the actual change is related to the PR title and its description.

M0_ENF_META is a part of m0_entity_flags enumeration. For some reason, in the idx-related part of Motr client we use index operation flags (oi->oi_flags) to understand if M0_ENF_META behavior is enabled. Now, we have a patch that changes one single place, so that entity flags (en_flags) are used there instead of operation flags. This raises the following questions:

1. How should M0_ENF_META flag be used for index-related Motr client operations? is there any specification in the code (a comment or a UT) where we tell that oi->oi_flags should be set?

2. Why did we use oi->oi_flags & M0_ENF_META for idx in the original code? obj-related part of the client uses the right entity flags (en_flags).

3. The patch replaces oi_flags with en_flags in one single place. What about the other places where oi_flags & M0_ENF_META is used? Do we need to fix them too?

**Comment by imvenkip** [comment]
> > > Could you please elaborate - why your change improves performance and on how much (if there are any measurements)?
> 
> > Added M0_ENF_META flag to identify metadata. Client(s3server) will store the pool version and layout type if the MO_ENF_META flag is set while index creation. Motr will store the pool version and layout type if the
> > M0_ENF_META flag is not set while index creation. Please check reference EOS-23040.
> 
> > EOS-22704 is addressing from Motr and S3. I collected stats with 1726 main build and with motor changes.
> > Please check in JIRA link to find small obj read performance improvement stats.
> 
> It is still not clear how the actual change is related to the PR title and its description.
> 
> M0_ENF_META is a part of m0_entity_flags enumeration. For some reason, in the idx-related part of Motr client we use index operation flags (oi->oi_flags) to understand if M0_ENF_META behavior is enabled. Now, we have a patch that changes one single place, so that entity flags (en_flags) are used there instead of operation flags. This raises the following questions:
> 
> 1. How should M0_ENF_META flag be used for index-related Motr client operations? is there any specification in the code (a comment or a UT) where we tell that oi->oi_flags should be set?
> 2. Why did we use oi->oi_flags & M0_ENF_META for idx in the original code? obj-related part of the client uses the right entity flags (en_flags).
> 3. The patch replaces oi_flags with en_flags in one single place. What about the other places where oi_flags & M0_ENF_META is used? Do we need to fix them too?

@andriytk @ivan-alekhin  It is a continuation of  Meta data optimization path. Could you check below design link? I believe, It will clarify relevant queries. 

https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/297862297/md-path+optimization+by+avoiding+internal+dix+layout+lookups

**Comment by andriytk** [comment]
@imvenkip people from Open Source community do **not** have access to our Confluence pages as well as to Jira tickets. So would it be possible to add some context and documentation about the problem into the description of this PR?

**Comment by imvenkip** [comment]
> @imvenkip people from Open Source community do **not** have access to our Confluence pages as well as to Jira tickets. So would it be possible to add some context and documentation about the problem into the description of this PR? cc @johnbent

Updated description in commit message.

**Pull request merged**